ee/hogai/graph/entity_search/toolkit.py

```diff
@@ -0,0 +1,208 @@
+import asyncio
+from collections import defaultdict
+from typing import TypedDict
+
+import yaml
+from posthoganalytics import capture_exception
+
+from posthog.api.search import EntityConfig, class_queryset
+from posthog.models import Action, Cohort, Dashboard, Experiment, FeatureFlag, Insight, Survey, Team, User
+from posthog.rbac.user_access_control import UserAccessControl
+from posthog.sync import database_sync_to_async
+
+from ee.hogai.utils.types.base import EntityType
+
+from .prompts import ENTITY_TYPE_SUMMARY_TEMPLATE, FOUND_ENTITIES_MESSAGE_TEMPLATE, HYPERLINK_USAGE_INSTRUCTIONS
+
+ENTITY_MAP: dict[str, EntityConfig] = {
+    "insight": {
+        "klass": Insight,
+        "search_fields": {"name": "A", "description": "C", "query_metadata": "B"},
+        "extra_fields": ["name", "description", "query", "query_metadata"],
+    },
+    "dashboard": {
+        "klass": Dashboard,
+        "search_fields": {"name": "A", "description": "C"},
+        "extra_fields": ["name", "description"],
+    },
+    "experiment": {
+        "klass": Experiment,
+        "search_fields": {"name": "A", "description": "C"},
+        "extra_fields": ["name", "description"],
+    },
+    "feature_flag": {"klass": FeatureFlag, "search_fields": {"key": "A", "name": "C"}, "extra_fields": ["key", "name"]},
+    "action": {
+        "klass": Action,
+        "search_fields": {"name": "A", "description": "C"},
+        "extra_fields": ["name", "description"],
+    },
+    "cohort": {
+        "klass": Cohort,
+        "search_fields": {"name": "A", "description": "C", "filters": "B"},
+        "extra_fields": ["name", "description", "filters"],
+    },
+    "survey": {
+        "klass": Survey,
+        "search_fields": {"name": "A", "description": "C"},
+        "extra_fields": ["name", "description"],
+    },
+}
+"""
+Map of entity names to their class, search_fields and extra_fields.
+
+The value in search_fields corresponds to the PostgreSQL weighting i.e. A, B, C or D.
+"""
+
+
+class EntitySearchTaskResult(TypedDict):
+    entity_type: str | EntityType
+    results: list[dict]
+    warning: str | None
+
+
+class EntitySearchToolkit:
+    MAX_ENTITY_RESULTS = 10
+    MAX_CONCURRENT_SEARCHES = 10
+
+    def __init__(self, team: Team, user: User):
+        self._team = team
+        self._user = user
+
+    @property
+    def user_access_control(self) -> UserAccessControl:
+        return UserAccessControl(user=self._user, team=self._team, organization_id=self._team.organization.id)
+
+    def build_url(self, entity_type: str | EntityType, result_id: str) -> str:
```

Comment:
I don't see it is used publicly:
```suggestion
    def _build_url(self, entity_type: str | EntityType, result_id: str) -> str:
```